### PR TITLE
146 extend the tests for multihop swap

### PR DESCRIPTION
**Pull request by gangov**

### CODE REVIEW

File: contracts/multihop/src/tests/swap.rs [hunk 1]
```diff
@@ -157,6 +157,482 @@ fn swap_three_equal_pools_no_fees() {
     assert_eq!(tokens[3].balance(&recipient), 50i128);
 }
 
+#[test]
+fn swap_single_pool_no_fees() {
+    let env = Env::default();
+
+    let admin = Address::random(&env);
+    let user = Address::random(&env);
```
Comment (ueco-jb) [inline contracts/multihop/src/tests/swap.rs:165]: You are using `user` account only for liquidity, this as well may be done by `admin`.

File: contracts/multihop/src/tests/swap.rs [hunk 2]
```diff
@@ -157,6 +157,482 @@ fn swap_three_equal_pools_no_fees() {
...
+    let mut token2 = deploy_token_contract(&env, &admin);
+
+    env.mock_all_auths();
+    env.budget().reset_unlimited();
+
+    let mut tokens = [&mut token1, &mut token2];
+    tokens.sort_by(|a, b| a.address.cmp(&b.address));
+
+    tokens[0].mint(&user, &10_000_000i128);
+    tokens[1].mint(&user, &10_000_000i128);
```
Comment (ueco-jb) [inline contracts/multihop/src/tests/swap.rs:177]: This can be first framework function.

File: contracts/multihop/src/tests/swap.rs [hunk 3]
```diff
@@ -157,6 +157,482 @@ fn swap_three_equal_pools_no_fees() {
...
+
+    // 1. deploy factory
+    let factory_addr = deploy_factory_contract(&env, admin.clone());
+    let factory_client = factory::Client::new(&env, &factory_addr);
+
+    factory_client.initialize(&admin.clone());
```
Comment (ueco-jb) [inline contracts/multihop/src/tests/swap.rs:183]: This will be a second.

File: contracts/multihop/src/tests/swap.rs [hunk 4]
```diff
@@ -157,6 +157,482 @@ fn swap_three_equal_pools_no_fees() {
...
+        let lp_client = lp_contract::Client::new(&env, &lp);
+        lp_client.provide_liquidity(
+            &user.clone(),
+            &Some(1_000_000i128),
+            &None,
+            &Some(1_000_000i128),
+            &None,
+            &None::<i64>,
+        );
+    }
```
Comment (ueco-jb) [inline contracts/multihop/src/tests/swap.rs:225]: And this should be a third one.
Here the problem is a query on factory returning an address of liquidity pool, but this can be easily resolved through #149.
Do #149 first, then previous call `factory_client.create_liquidity_pool(&first_lp_init_info);` will return an address to use in creating a client.

File: contracts/multihop/src/tests/swap.rs [hunk 5]
```diff
@@ -157,6 +157,482 @@ fn swap_three_equal_pools_no_fees() {
...
+        ask_asset: tokens[1].address.clone(),
+    };
+
+    let operations = vec![&env, swap1];
+
+    multihop.swap(&recipient, &operations, &50i128);
+
+    // 5. check if it goes according to plan
+    assert_eq!(tokens[0].balance(&recipient), 0i128);
+    assert_eq!(tokens[1].balance(&recipient), 43i128);
```
Comment (ueco-jb) [inline contracts/multihop/src/tests/swap.rs:333]: Do decided to choose arbitrary difficult to calculate fee (1500 bps which is 15%) and very small amount for swap.
Additinally you didn't provide math for result and just asserted it.
This is a very bad practice. Why do we have tests, if it just asserts what we already know? 
Add comments such as:
```
// 1:1 pool of 1M tokens
// Swapping 50 tokens with 15% swap fee
// 0.85 * 50 = 42.5 ~= 43 (rounded up)

**Review by ueco-jb — COMMENTED** [review]
Those few comments really apply for all testcases.

File: contracts/multihop/src/tests/swap.rs [hunk 6]
```diff
@@ -178,3 +580,41 @@ fn swap_panics_with_no_operations() {
...
+        true => {
+            let mut i = 1i128;
+            tokens.iter().for_each(|mut t| {
+                t.mint(&admin, &(i * 10_000_000i128));
+                i += 1;
+            });
+        }
+        false => {
+            tokens.iter().for_each(|t| t.mint(&admin, &10_000_000i128));
+        }
```
Comment (ueco-jb) [inline contracts/multihop/src/tests/swap.rs:608]: I specifically didn't mention token generation in my comments regarding making helper functions.
1) Sorting doesn't make sense. Tokens needs to be sorted just before introducing as a parameters for liquidity pools creation.
2) Different token ratio as a boolean doesn't make sense. We want to have a control over ratios.
3) Minting tokens doesn't have anything to do with token ratio. It matters during initial liquidity providing, which is I wrote on mattermost:
> let pool = deploy_and_initialize_lp(token_a, token_a_amount, token_b, token_b_amount, Option<fees>, Option<slippage>)

Because that's where using liquidity and token amounts matters.

File: contracts/multihop/src/tests/swap.rs [hunk 7]
```diff
@@ -13,27 +13,11 @@ fn swap_three_equal_pools_no_fees() {
...
-    let mut token1 = deploy_token_contract(&env, &admin);
-    let mut token2 = deploy_token_contract(&env, &admin);
-    let mut token3 = deploy_token_contract(&env, &admin);
-    let mut token4 = deploy_token_contract(&env, &admin);
-
-    env.mock_all_auths();
-    env.budget().reset_unlimited();
-
-    let mut tokens = [&mut token1, &mut token2, &mut token3, &mut token4];
-    tokens.sort_by(|a, b| a.address.cmp(&b.address));
```
Comment (ueco-jb) [inline contracts/multihop/src/tests/swap.rs:26]: Actually it doesn't make sense to sort those tokens here, it needs to be done just in a helper of creating liquidity pool. That's where it matters.

**Review by ueco-jb — COMMENTED** [review]
I guess all in all you could create a small function:
```rust
fn deploy_and_mint_tokens(env: &Env, admin: Address, amount: i128) -> token_contract::Client {
    let token = deploy_token_contract(&env, &admin);
    token.mint(&admin, &amount);
    token
}
```

File: contracts/multihop/src/tests/swap.rs [hunk 8]
```diff
@@ -649,3 +580,41 @@ fn swap_panics_with_no_operations() {
...
+    tokens
+}
+
+fn get_factory(env: &Env, admin: Address) -> factory::Client {
+    let factory_addr = deploy_factory_contract(&env, admin.clone());
+    let factory_client = factory::Client::new(&env, &factory_addr);
+
+    factory_client.initialize(&admin.clone());
+    factory_client
+}
```
Comment (ueco-jb) [inline contracts/multihop/src/tests/swap.rs:620]: This one is good.

File: contracts/multihop/src/tests/swap.rs (hunk 6 above)
Comment (gangov) [reply contracts/multihop/src/tests/swap.rs:608]: the highlighted code is gone and replaced with the example from above

File: contracts/multihop/src/tests/swap.rs (hunk 3 above)
Comment (gangov) [reply contracts/multihop/src/tests/swap.rs]: fixed with `deploy_and_mint_tokens` as per example

File: contracts/multihop/src/tests/swap.rs (hunk 2 above)
Comment (gangov) [reply contracts/multihop/src/tests/swap.rs]: fixed as per `deploy_and_mint_tokens` example below

File: contracts/multihop/src/tests/swap.rs (hunk 7 above)
Comment (ueco-jb) [reply contracts/multihop/src/tests/swap.rs:26]: Done.

File: contracts/multihop/src/tests/swap.rs (hunk 4 above)
Comment (ueco-jb) [reply contracts/multihop/src/tests/swap.rs]: Done.

File: contracts/multihop/src/contract.rs [hunk 9]
```diff
@@ -67,7 +67,7 @@ impl MultihopTrait for Multihop {
                 &op.offer_asset,
                 &next_offer_amount,
                 &None::<i64>,
-                &Some(5000i64),
+                &None::<i64>,
```
Comment (ueco-jb) [inline contracts/multihop/src/contract.rs:70]: Wait, where does this change comes from? If anything we should add this parameter to this message, not enforce `None`. And it should be in a separate PR.

Comment (gangov) [reply contracts/multihop/src/contract.rs:70]: thought it won't hurt us, reverting now

File: contracts/multihop/src/tests/swap.rs (hunk 5 above)
Comment (gangov) [reply contracts/multihop/src/tests/swap.rs]: I believe this is now properly asserted and proved with comments above the assertions

File: contracts/multihop/src/tests/swap.rs [hunk 10]
```diff
@@ -3,158 +3,363 @@ use crate::storage::Swap;
...
-        swap_fee_bps: 0,
-        token_init_info: second_token_init_info,
-        stake_init_info: second_stake_init_info,
+    let operations = vec![&env, swap1];
+
+    multihop.swap(&recipient, &operations, &1);
+
+    // 5. check if it goes according to plan
+    assert_eq!(token1.balance(&recipient), 49i128); // -1 token0
+    assert_eq!(token2.balance(&recipient), 1i128); // +1 token1
```
Comment (ueco-jb) [inline contracts/multihop/src/tests/swap.rs:132]: Increase those numbers, swapping of 1 doesn't really show anything.
```suggestion
    multihop.swap(&recipient, &operations, &50);

    // 5. check if it goes according to plan
    assert_eq!(token1.balance(&recipient), 0i128);
    assert_eq!(token2.balance(&recipient), 50i128);
```

File: contracts/multihop/src/tests/swap.rs [hunk 11]
```diff
@@ -3,158 +3,363 @@ use crate::storage::Swap;
...
+    let operations = vec![&env, swap1];
+
+    multihop.swap(&recipient, &operations, &300i128);
+
+    // 5. check if it goes according to plan
+    // 1000 tokens initially
+    // swap 300 from token0 to token1 with 2000 bps (20%)
+    // tokens1 will be 240
+    assert_eq!(token1.balance(&recipient), 700i128);
+    assert_eq!(token2.balance(&recipient), 240i128);
```
Comment (ueco-jb) [inline contracts/multihop/src/tests/swap.rs:181]: Very good. :+1:

File: contracts/multihop/src/tests/swap.rs [hunk 12]
```diff
@@ -3,158 +3,363 @@ use crate::storage::Swap;
...
+    let multihop = deploy_multihop_contract(&env, admin.clone(), &factory_client.address);
     let recipient = Address::random(&env);
-    tokens[0].mint(&recipient, &50i128);
-    assert_eq!(tokens[0].balance(&recipient), 50i128);
-    assert_eq!(tokens[3].balance(&recipient), 0i128);
+    token1.mint(&recipient, &10_000i128);
+    assert_eq!(token1.balance(&recipient), 10_000i128);
+    assert_eq!(token2.balance(&recipient), 0i128);
+    assert_eq!(token3.balance(&recipient), 0i128);
+    assert_eq!(token4.balance(&recipient), 0i128);
```
Comment (ueco-jb) [inline contracts/multihop/src/tests/swap.rs:321]: Let's test what needs to be tested, we don't need to check that tokens that weren't minted before aren't there.

```suggestion
```

File: contracts/multihop/src/tests/swap.rs [hunk 13]
```diff
@@ -3,158 +3,363 @@ use crate::storage::Swap;
...
+
+    // token3 => token4
+    // (7985 * 3_000_000) / (7985 + 3_000_000)
+    // 23_955_000_000 / 3_007_985
+    // 7963.80301099
+    // 7964 - 10% = 7168
+    assert_eq!(token1.balance(&recipient), 0i128);
+    assert_eq!(token2.balance(&recipient), 0i128);
+    assert_eq!(token3.balance(&recipient), 0i128);
+    assert_eq!(token4.balance(&recipient), 7_168i128);
```
Comment (ueco-jb) [inline contracts/multihop/src/tests/swap.rs:362]: Awesome. :+1:

**Review by ueco-jb — COMMENTED** [review]
Adjust those couple things, other then that it's good.

**Comment by gangov** [comment]
@gangov: note to self, add another set of assertions that verifies the amount of tokens in between swaps

File: contracts/multihop/src/tests/swap.rs [hunk 14]
```diff
@@ -3,158 +3,360 @@ use crate::storage::Swap;
...
 
-    multihop.swap(&recipient, &operations, &50i128);
+    multihop.swap(&recipient, &operations, &10_000i128);
 
-    // 5. check if it goes according to plan
-    assert_eq!(tokens[0].balance(&recipient), 0i128);
-    assert_eq!(tokens[3].balance(&recipient), 50i128);
+    // we start swapping 10_000 tokens
+
+    // token1 => token2
```
Comment (gangov) [inline contracts/multihop/src/tests/swap.rs:339]: add another set of assertions that verifies the amount of tokens in between swaps in addition to the comments

**Review by ueco-jb — APPROVED** [review]
LGTM
